### PR TITLE
Add the bootloader stub.

### PR DESCRIPTION
**Pull request by nomis52**
This correctly enumerates as a DFU device.

### CODE REVIEW

**Comment by peternewman** [comment]
Fix Travis

**Comment by nomis52** [comment]
Fixed BTW

File: Bootloader/firmware/src/app.c [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+          setupPacket->DataDir == USB_SETUP_REQUEST_RECIPIENT_INTERFACE &&
+          setupPacket->wIndex == USB_DFU_INTERFACE_INDEX) {
+        HandleDFUEvent(setupPacket);
+      } else if (setupPacket->bRequest == USB_REQUEST_SET_INTERFACE) {
+        /* If we have got the SET_INTERFACE request, we just acknowledge
+         for now. This demo has only one alternate setting which is already
+         active. */
+        USB_DEVICE_ControlStatus(g_app.usb_device,
+                                 USB_DEVICE_CONTROL_STATUS_OK);
+        /*
```
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: Remove?

Comment (peternewman) [inline Bootloader/firmware/src/app.c]: Or TODO

File: Bootloader/firmware/src/app.c [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+    // These events are not used.
+    case USB_DEVICE_EVENT_ENDPOINT_READ_COMPLETE:
+    case USB_DEVICE_EVENT_ENDPOINT_WRITE_COMPLETE:
+    case USB_DEVICE_EVENT_RESUMED:
+    case USB_DEVICE_EVENT_ERROR:
+    default:
+      break;
+  }
+  (void) context;
```
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: UNUSED macro instead

File: Bootloader/firmware/src/app.c [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+void APP_Tasks(void) {
+  switch (g_app.state) {
+    case APP_STATE_INIT:
+      // Try to open the device layer.
+      g_app.usb_device = USB_DEVICE_Open(USB_DEVICE_INDEX_0,
+                                         DRV_IO_INTENT_READWRITE);
+      if (g_app.usb_device != USB_DEVICE_HANDLE_INVALID) {
+        // Register a callback with device layer to get event notification
+        // for end point 0.
```
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: SPaG endpoint

File: Bootloader/firmware/src/app.c [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        USB_DEVICE_EventHandlerSet(g_app.usb_device, USBEventHandler, 0);
```
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: Make 0 a constant

File: Bootloader/firmware/src/app.c [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+      }
+      break;
+
+    case APP_STATE_MAIN_TASK:
+      if (!g_app.is_configured) {
+        /* This means the device got deconfigured. Change the
+         * application state back to waiting for configuration. */
+        g_app.state = APP_STATE_WAIT_FOR_CONFIGURATION;
+      } else {
+        // Do something
```
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: TODO

File: Bootloader/firmware/src/app.h [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ * app.h
+ * Copyright (C) 2015 Simon Newton
+ */
+
+#ifndef _APP_H
```
Comment (peternewman) [inline Bootloader/firmware/src/app.h]: Shouldn't these be the full path?

**Comment by peternewman** [comment]
Generally LGTM

File: Bootloader/firmware/src/app.c (hunk 1 above)
Comment (nomis52) [inline Bootloader/firmware/src/app.c]: This is added back in the next PR

File: Bootloader/firmware/src/app.c (hunk 2 above)
Comment (nomis52) [inline Bootloader/firmware/src/app.c]: Done in the next PR

File: Bootloader/firmware/src/app.c (hunk 3 above)
Comment (nomis52) [inline Bootloader/firmware/src/app.c]: Fixed

File: Bootloader/firmware/src/app.c (hunk 4 above)
Comment (nomis52) [inline Bootloader/firmware/src/app.c]: Switched to NULL

File: Bootloader/firmware/src/app.c (hunk 5 above)
Comment (nomis52) [inline Bootloader/firmware/src/app.c]: Added in the next PR

File: Bootloader/firmware/src/app.h (hunk 6 above)
Comment (nomis52) [inline Bootloader/firmware/src/app.h]: Fixed

File: Bootloader/firmware/src/app.c (hunk 1 above)
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: Okay

File: Bootloader/firmware/src/app.c (hunk 4 above)
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: Ah, I'd assumed it was the endpoint number.

**Comment by peternewman** [comment]
LGTM, will check the next PR.

**Comment by peternewman** [comment]
Next PR LGTM

**Pull request merged**